crates/bevy_asset/src/path.rs

```diff
@@ -141,47 +155,100 @@ impl<'a> AssetPath<'a> {
     }
 }
 
-impl<'a> From<&'a str> for AssetPath<'a> {
-    fn from(asset_path: &'a str) -> Self {
-        let mut parts = asset_path.splitn(2, '#');
-        let path = Path::new(parts.next().expect("Path must be set."));
-        let label = parts.next();
-        AssetPath {
-            path: Cow::Borrowed(path),
-            label: label.map(Cow::Borrowed),
-        }
+impl From<&'static str> for AssetPath<'static> {
+    #[inline]
+    fn from(asset_path: &'static str) -> Self {
+        AssetPath::new(asset_path)
     }
 }
 
 impl<'a> From<&'a String> for AssetPath<'a> {
+    #[inline]
     fn from(asset_path: &'a String) -> Self {
-        asset_path.as_str().into()
+        AssetPath::new(asset_path.as_str())
     }
 }
 
-impl<'a> From<&'a Path> for AssetPath<'a> {
-    fn from(path: &'a Path) -> Self {
-        AssetPath {
-            path: Cow::Borrowed(path),
+impl From<String> for AssetPath<'static> {
+    #[inline]
+    fn from(asset_path: String) -> Self {
+        AssetPath::new(asset_path.as_str()).into_owned()
+    }
+}
+
+impl From<&'static Path> for AssetPath<'static> {
+    #[inline]
+    fn from(path: &'static Path) -> Self {
+        Self {
+            path: path.into(),
             label: None,
         }
     }
 }
 
-impl<'a> From<PathBuf> for AssetPath<'a> {
+impl From<PathBuf> for AssetPath<'static> {
+    #[inline]
     fn from(path: PathBuf) -> Self {
-        AssetPath {
-            path: Cow::Owned(path),
+        Self {
+            path: path.into(),
             label: None,
         }
     }
 }
 
+impl<'a, 'b> From<&'a AssetPath<'b>> for AssetPath<'b> {
+    fn from(value: &'a AssetPath<'b>) -> Self {
+        value.clone()
+    }
+}
+
 impl<'a> From<AssetPath<'a>> for PathBuf {
-    fn from(path: AssetPath<'a>) -> Self {
-        match path.path {
-            Cow::Borrowed(borrowed) => borrowed.to_owned(),
-            Cow::Owned(owned) => owned,
-        }
+    fn from(value: AssetPath<'a>) -> Self {
+        value.path().to_path_buf()
+    }
+}
+
+impl<'a> Serialize for AssetPath<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut state = serializer.serialize_tuple_struct("AssetPath", 1)?;
+        let string = self.to_string();
+        state.serialize_field(&string)?;
+        state.end()
```

Comment:
This should probably be added to the Migration Guide as well.

Reply:
Good call!